src/ItemTreeTraversal/BaseItemTreeTraversal.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Rbac\Cycle\ItemTreeTraversal;
+
+use Cycle\Database\DatabaseInterface;
+
+/**
+ * Base class for all RBAC item tree traversal strategies containing required dependencies.
+ *
+ * @internal
+ */
+abstract class BaseItemTreeTraversal
```

Comment:
Is there a need for this class and inheritance? There's no common logic involved so I'd move that to actual strategies.

Reply:
There is common logic with CTE:

> The base queries are unified as much possible to work for all RDBMS supported by Cycle with minimal differences.

The only differences are specifying `WITH` expression in SQL Server and casting "name" column type in MySQL. They are covered using additional methods. This is better than duplicating the same query 4 times in my opinion.

Also it's needed to declare required dependencies in constructor, otherwise there will be 5 places with copy pasted code.

Reply:
1. There's no common interface implemented by this class.
2. There are no check for `BaseItemTreeTraversal` anywhere.
3. There's no logic inside except constructor.

Reply:
Maybe be convert it to trait then?